### DISCUSSION TITLE
Making AppImages work on any linux system

### DISCUSSION THREAD

**Discussion by Samueru-sama** [ivan-hc/AM#1120] [Show and tell]
I guess this is a follow up of #369

Lately I've been updating all my appimages to work on any linux system by deploying all the libs including the dynamic linker without using containers or similar tricks, which lets them work on any linux system, be it very old or a musl system.

I originally used go-appimage's deploy everything mode, but I just moved to either doing it manually or using [sharun](https://github.com/VHSgunzo/sharun).

The following appimages should work on any distro: 

https://github.com/Samueru-sama/GIMP-AppImage
https://github.com/Samueru-sama/mpv-AppImage
https://github.com/Samueru-sama/OBS-Studio-AppImage
https://github.com/Samueru-sama/DeaDBeeF-AppImage
https://github.com/Samueru-sama/puddletag-AppImage
https://github.com/Samueru-sama/strawberry-AppImage
https://github.com/Samueru-sama/pavucontrol-qt-AppImage
https://github.com/Samueru-sama/SpeedCrunch-AppImage
https://github.com/Samueru-sama/android-tools-AppImage
https://github.com/Samueru-sama/polybar-AppImage
https://github.com/Samueru-sama/rofi-AppImage

I have also made PRs on other repos to have such appimages: 

https://github.com/AlexanderP/tesseract-appimage/pull/4
https://github.com/Syllo/nvtop/pull/324 (Next release with the changes hasn't been made)
https://github.com/dunst-project/dunst/pull/1391 (Still open)
https://github.com/psadi/ghostty-appimage/pull/2

--------------------------

CC @Drsheppard01

**Comment by Drsheppard01** [comment]
Hi guys. I can test it in my system (Chimera Linux, musl, chimerautils)

- [x] GIMP (work pretty)
- [x] mpv
- [x] OBS
- [x] DeaDBeeF
- [x] Puddletag
- [x] Strawberry
- [x] pavucontrol-qt
- [x] Speed-Crunch
- [x] android-tools
- [x] polybar (included config not uploaded)
- [x] rofi (can't exit from rofi, looks like touchpad or keyboard isn't initialized)

btw, need help to improve gyroflow build scripts https://github.com/gyroflow/gyroflow

**Reply by Samueru-sama** [reply]
>polybar (included config not uploaded)

Looks like polybar expects a default config file on `/etc` when the user doesn't have one. Should be an easy fix with some hook checking if there isnt a user config file copies the one in the AppDir to that location.

> rofi (can't exit from rofi, looks like touchpad or keyboard isn't initialized)

I tested the appimage on a void musl vm and I can't replicate this issue. 🤔

>btw, need help to improve gyroflow build scripts https://github.com/gyroflow/gyroflow

This? https://github.com/gyroflow/gyroflow/blob/373dd2bcaf4c5a16ec5e450d40a4f7adf732cb4a/_deployment/linux/AppImageBuilder.yml

All of that has to go xd, the old appimagebuilder scripts even use an [AppRun](https://github.com/AppImageCrafters/AppRun) that has a dependency to glibc.

sharun actually has the `.env`capability of that AppRun as well, plus it already sets several [env variables](https://github.com/VHSgunzo/sharun/blob/main/src/main.rs#L382-L522) for certain locations that need them already.

sharun also hardlinks itself to the binaries being ran, which means this issue with [go-appimage](https://github.com/probonopd/go-appimage/issues/49) is avoided.

**Reply by Samueru-sama** [reply]
> btw, need help to improve gyroflow build scripts https://github.com/gyroflow/gyroflow

In any case I'm on the loomix telegram channel if you need help, you can find the link on the readme of this [repo](https://github.com/Azathothas/Toolpacks)

**Reply by Samueru-sama** [reply]
> polybar (included config not uploaded)

Alright this should work now, I did the old school trick of [patching the binary](https://github.com/Samueru-sama/polybar-AppImage/blob/main/polybar-AppImage.sh#L31-L33) instead of having a check and copying the config in the AppRun.

**Comment by Samueru-sama** [comment]
Update: I updated the GIMP and OBS appimages to use this [appimage runtime](https://github.com/VHSgunzo/uruntime) that works with dwarfs, so now they are a bit smaller as result.

**Comment by jcelerier** [comment]
I'm curious, how does that work for opengl / libGL.so ? As you cannot ship multiple libGL.so but that's what is needed for nvidia vs non-nvidia

**Reply by Samueru-sama** [reply]
> I'm curious, how does that work for opengl / libGL.so ? As you cannot ship multiple libGL.so but that's what is needed for nvidia vs non-nvidia

Good question, not sure. 

Do you have nvidia? I recently had this issue: https://github.com/pkgforge-dev/PPSSPP-AppImage/issues/7

But it ended up being the user having a too old GPU since the problem is present with other methods like flatpak. 

I already bundle nvk libs on my appimages, but as far as I know that's vulkan and not opengl. I think the proprietary nvidia driver would be dlopened on host, which means I wouldn't work on appimages that ship musl, but I don't think I have an appmage that uses the video drivers that ships musl though.

**Reply by jcelerier** [reply]
nvidia driver doesn't work on musl systems at all anyways, it explicitly requires glibc

**Reply by Samueru-sama** [reply]
>nvidia driver doesn't work on musl systems at all anyways, it explicitly requires glibc

Yeah that's what I said at the end 

"I think the proprietary nvidia driver would be dlopened on host, which means I wouldn't work on appimages that ship musl, but I don't think I have an appmage that uses the video drivers that ships musl though."

**Reply by Samueru-sama** [reply]
> I'm curious, how does that work for opengl / libGL.so ? As you cannot ship multiple libGL.so but that's what is needed for nvidia vs non-nvidia

Just dusted out a GTX1060 that I have and carried some tests. 

I'm able to use the nvidia driver on the AppImages of [desmume](https://github.com/pkgforge-dev/DeSmuME-AppImage) (opengl) and [ppsspp](https://github.com/pkgforge-dev/PPSSPP-AppImage) (vulkan).

I was not able to get the driver to be used in this [appimage](https://github.com/prusa3d/PrusaSlicer/issues/13653#issuecomment-2579047665) of prusaslicer though, however I got someone else with an RTX3060 to test it and was told that it did use the driver, but I need to make sure how they knew the driver was being used.

**Reply by Samueru-sama** [reply]
Update: I was able to solve the issue with PrusaSlicer now it correctly uses the proprietary nvidia driver from the host.

It ended up being a bug in the old version of mesa that was bundled on that appimage that prevented it from loading the nvidia driver lol

**Comment by vishnu350** [comment]
Hi @Samueru-sama, does Anylinux do the [magic bytes patch](https://github.com/AppImage/AppImageKit/issues/828) thing so that it works within Docker/Podman?

**Reply by Samueru-sama** [reply]
> Hi @Samueru-sama, does Anylinux do the [magic bytes patch](https://github.com/AppImage/AppImageKit/issues/828) thing so that it works within Docker/Podman?

Nope, doing that trick would mean the appimage would stop being an appimage.

The type2 spec says that the first 10 bytes of the AppImage must contain `ELF....AI`

This is what your file manager uses to know that the file is an appimage, what lets your appimage thumbnailer work on that appimage, what lets `appimageupdatetool` update that appimage, etc ,etc

---

With that said I don't have issues running appimages in podman, so no idea what the problem is exactly.

**Reply by vishnu350** [reply]
> With that said I don't have issues running appimages in podman, so no idea what the problem is exactly.

With libfuse installed and `--device /dev/fuse --cap-add SYS_ADMIN` setup, I don't have issues either. So I was wondering if you guys did some kind of fix, since their issue is still open.

**Reply by Samueru-sama** [reply]
> With libfuse installed and `--device /dev/fuse --cap-add SYS_ADMIN` setup, I don't have issues either. So I was wondering if you guys did some kind of fix, since their issue is still open.

We do not need FUSE, because the uruntime is able to use mountnamespaces, see: 

* https://github.com/VHSgunzo/uruntime/issues/18
* https://github.com/VHSgunzo/uruntime?tab=readme-ov-file#built-in-configuration

The type2-runtime has this PR pending adding the same feature: 

* https://github.com/AppImage/type2-runtime/pull/137#event-20651671670

**Reply by vishnu350** [reply]
Thanks @Samueru-sama, anylinux is awesome that it works even for really old stuff. I was looking into tests for AM (to be run in podman) for a PR, and some appimages still need fuse support though.

BTW, how do I list the smallest sized appimages from pkgforge? I want to randomly pick a few.

**Reply by Samueru-sama** [reply]
> BTW, how do I list the smallest sized appimages from pkgforge? I want to randomly pick a few.

I don't think you can, here is a few that are small: 

* https://github.com/pkgforge-dev/xeyes-AppImage
* https://github.com/pkgforge-dev/rofi-AppImage
* https://github.com/pkgforge-dev/alacritty-AppImage
* https://github.com/pkgforge-dev/foot-AppImage
* https://github.com/pkgforge-dev/pavucontrol-qt-AppImage
